src/tunnel.cc

```diff
@@ -1307,37 +1298,74 @@ TunnelStateData::saveError(ErrorState *error)
 {
     debugs(26, 4, savedError << " ? " << error);
     assert(error);
+    // nobody saves finalError (or makes finalError equal to savedError)
+    assert(!finalError || (finalError != error && finalError != savedError));
+    if (savedError == error)
+        return;
     delete savedError; // may be nil
     savedError = error;
 }
 
-/// Starts sending the given error message to the client, leading to the
-/// eventual transaction termination. Call with savedError to send savedError.
+/// Makes sure the transaction is terminated, immediately (if possible) or after
+/// reporting the previously saved error to the client and/or finishing in-progress
+/// I/Os (as needed).
+/// All callers guarantee savedError existence, which becomes the final error during
+/// the first call. Safe to call multiple times.
 void
-TunnelStateData::sendError(ErrorState *finalError, const char *reason)
+TunnelStateData::sendErrorAndDestroy(const char *reason)
```

Comment:
What do you think about renaming sendErrorAndDestroy() to terminateWithSavedError() or perhaps destroyWithSavedError()? It will highlight the caller responsibility to ensure than some error was saved prior to calling this method and possibly defend against "but you are not always sending the error!" attack by those who interpret the method name very literally.